planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+to publish to an event bus. _This way lies madness._ We ended up implementing the same algorithm that NServiceBus uses to "guarantee" eventual message 
+publishing after saving to the eventstore. This algorithm works, but its not the best performance-wise and it in theory requires infinite retries in order for
+the guarantee to actually be a guarantee. Right now we are where we are and reimplementing this part in our homegrown lib is not really on the table right
+now. But if I had to implement this again knowing what I do today I would 100% certainly implement publishing to the message bus as an async projection.
+
+Running projections off the message bus was also a thing that we tried doing for a while. It has some good parts, but also some bad parts. The main issue
+is of course that you suddenly end up having to handle things like out-of-order messages, duplicated messages, not being sure of where you are in the stream
+etc. Its quite a lot to keep in mind, but for some cases its quite useful. But complexity immediately rises, especially in writing the projections, and lets not
+even talk about event replay. I saw that you were discussing allowing projections from a message bus, and it would be a great feature to have but be prepared
+that in order to support the same type of projections as you can do today is gonna require quite a bit of infrastructure. But if you can pull it off, Id certainly be happy! ;)
+
+
+# Immutable objects
+
+Honestly, you knew this were coming. But hear me out.
+
+## Immutable events
+Events should be immutable, thats just a basic fact of eventsourcing. But due to serialization woes etc the common implementation of events are with 
+`public string Fruit { get; set; }` because thats just life in C# world. Now, this "gentleman immutability" "works" but it is...not great?
+
+### It aint immutable
+Sure, if you go assign stuff to it and end up in trouble because of it thats your own fault. But I am also a firm believer in the principle of least surprise, if you shouldnt 
+set a property there shouldnt be a setter for it. If a method shouldnt be in the public interface of a class then make it private or internal.
+
+### F# support
+Mutable bags of data like this is highly non-idiomatic in F#, and key to providing good F# support is going to be in part allowing for better support for 
+immutable object. The normal modus operandi there would be to implement an event as a record type (which in turn is probably wrapped in a DU, but lets table that discussion for right now)
+
+```fsharp
+type MyEvent = {
+  fruit : string
+  numberOfFruitsBought : int
+}
+```
+
+This event is immutable. By adding the `[<CLIMutable>]` attribute on it we can get public setters generated on it as well that cant be used from F#, solving some
+issues with immutability and serialization but at the cost of your immutable object...not actually being immutable anymore.
```

Comment:
If the immutable types are not working well with Marten then we should work on making that happen ;)